h2: Make requests walk away from the waiting list

Dridi opened the pull request:
When a session goes to the cleanup step, it will indefinitely try to
kill all the streams until those stuck in a waiting list reembark a
thread. This can lead to clients timing out, closing the h2 session,
and opening a new h2 session for the exact same requests. This can
artifically inflate the number of dead tasks in a waiting list with
generous backend timeouts.

To avoid that, an h2 stream will keep a list of requests currently
in a waiting list and make them walk away quietly during cleanup.
An h2 stream would normally only have one request running at any
time but this may no longer be true with modules like vmod_pesi.

---

See https://varnish-cache.org/lists/pipermail/varnish-misc/2022-July/027160.html

This is equivalent to what we have in Varnish Enterprise that I
mentioned in this thread. It does not solve the problem described
in the email thread, because we have no path to walk away from
an HTTP/1 request (pun intended).

I believe we could potentially solve the HTTP/1 case if we involved
a waiter that would only care about closed connections, ignoring
incoming data that I believe trigger a wake up today.

I am convinced that I managed a better design than the original
attempt, and I'm looking forward to reviews. Unfortunately, I will
miss the next two bugwash meetings.

nigoroll commented:
generally in favor, have not looked closed yet, still :(

mbgrydeland commented on bin/varnishd/cache/cache.h:
```diff
@@ -466,11 +466,13 @@ struct req {
 	void			*transport_priv;
 
 	VTAILQ_ENTRY(req)	w_list;
+	VTAILQ_ENTRY(req)	t_list;
```
This feels like quite the waste of memory adding 24 bytes for the list entry and the objhead reference flag on struct req. I think I prefer the original way of doing the workspace allocation, perhaps replaced/expanded with a fallback malloc.

Dridi replied:
I just remembered why I didn't use the workspace: it is reserved for vary processing.

Dridi replied:
I got some results on the generalization side, but didn't manage to without allocation space somewhere. Since workspace is reserved during waiting list operations, it is either a heap allocation or preallocated storage in struct req (current solution). Another possibility could be to opportunistically allocate an "entry" for transport tracking on the workspace before the vary reservation, if the top transport has waiting list methods.

To be continued.

Dridi replied:
Considering that eventually both h2 and http/1 (100% of client transports) will track requests and sub-requests in waiting lists, it would be much simpler to go with the current design and have the tracking state directly in struct req.

mbgrydeland commented on bin/varnishd/http2/cache_http2_session.c:
```diff
@@ -444,17 +445,59 @@ h2_new_session(struct worker *wrk, void *arg)
+
+	Lck_Lock(&r2->h2sess->sess->mtx);
+	AZ(req->transport_objhead);
+	req->transport_objhead = oh;
```
The variable `req->transport_objhead` should be renamed `req->toptransport_objhead`, to reflect that it is "owned" by the top transport. As it is it would be trouble if e.g. esi_deliver wanted to use it. Same argument goes for `t_list`.

Dridi replied:
Or, maybe we could move `req::t_list` and `req::transport_objhead` to `reqtop::w_list` and `reqtop::hash_objhead`?

Dridi commented on bin/varnishd/cache/cache_hash.c:
```diff
@@ -355,6 +355,63 @@ hsh_insert_busyobj(const struct worker *wrk, struct objhead *oh)
+	req->waitinglist = 1;
+
+	if (req->top->topreq->transport->waitlist != NULL)
+		req->top->topreq->transport->waitlist(wrk, req);
```
In the commit where it is introduced, there is a comment stating that it is only for `req_top`'s transport. The reason for the lack of symmetry is that the "current" transport action is the same for all requests: go into the waiting list.

Without this patch series there is already a reembark asymmetry.

What we have done downstream is less idea in my opinion, because we made each transport responsible for checking whether there is a top `waitlist` callback to signal instead of centralizing this check like in the current iteration.

We could have a distinction between the sole `reembark` callback and two separate `top_waitlist` and `top_reembark`. I didn't because I felt it was too heavy-handed.

mbgrydeland replied:
> We could have a distinction between the sole `reembark` callback and two separate `top_waitlist` and `top_reembark`. I didn't because I felt it was too heavy-handed.

I think I prefer that option.

Dridi replied:
Done.

Dridi commented on bin/varnishd/cache/cache_req_fsm.c:
```diff
@@ -561,6 +561,21 @@ cnt_lookup(struct worker *wrk, struct req *req)
+		AZ(req->stale_oc);
+		VSLb_ts_req(req, "Waitinglist", W_TIM_real(wrk));
+		VSLb(req->vsl, SLT_Error, "The client is going away");
+		if (req->esi_level > 0) {
```
First, we can add a dedicated `HSH_TIMEOUT`. Otherwise I can use the same argument for the error message preventing us to use this branch for future timeouts. I think we will be fine if we dedicated `HSH_WALKAWAY` to the case where the client went away while a request was in a waiting list.

For the timeout case, we  probably want to have different semantics than plainly ending the transaction. Maybe a waiting list timeout would be a deferred `hash_ignore_busy`?

I must admit I don't remember precisely why we did that downstream. I think we needed this to interrupt ESI delivery and prevent it from recursing as the client is gone anyway.

Dridi replied:
There's actually a comment a few lines below explaining why we fail sub-requests:

```c
/* NB: Interrupt delivery upwards to avoid engaging
 * new sub-request tasks.
 */
```

Granted, it doesn't give many details.

Dridi commented:
Rebased against current master to solve the trivial merge conflict. I switched to more explicit `tansport::top_*()` methods as suggested. Not addressing other review items, yet.

Dridi commented:
Closing for now because #3992 will pull the objhead rug pretty hard.
